File: practice/graphs/open_the_lock_v1.py
from typing import List
from collections import deque

class Solution:
    def openLock(self, deadends: List[str], target: str) -> int:
        def nextTurns(s):
            N = len(s)
            res = list()
            for i in range(N):
                tmp = s[:i] + ('0' if s[i] == '9' else str(int(s[i])+1)) + s[i+1:]
                res.append(tmp)
                tmp = s[:i] + ('9' if s[i] == '0' else str(int(s[i])-1)) + s[i+1:]
                res.append(tmp)
            return res
        # Important note - converting the list into a set cuts down the runtime to a tenth.
        # With deadends list, runtime = 6751 ms.
        # With dead set, runtime = 662 ms.
        dead = set(deadends)
        dq = deque()
        seen = set()
        start = '0000'
        if start in dead:
            return -1
        dq.append(start)
        seen.add(start)
        steps = 0
        while dq:
            size = len(dq)
            for _ in range(size):
                cur = dq[0]
                if cur == target:
                    return steps
                for t in nextTurns(cur):
                    if t not in dead and t not in seen:
                        seen.add(t)
                        dq.append(t)
                dq.popleft()
            steps += 1
        return -1

# Main section
for deadends, target in [
                           (['0201','0101','0102','1212','2002'], '0202'),
                           (['8888'], '0009'),
                           (['8887','8889','8878','8898','8788','8988','7888','9888'], '8888'),
                           (['0000'], '9999'),
                           (['8888'], '9999'),
                        ]:
    print(f'deadends, target = {deadends}, {target}')
    sol = Solution()
    r = sol.openLock(deadends, target)
    print(f'r = {r}')
    print('================')

  

File: algorithms/medium/open_the_lock_v1.go
package main

import "fmt"

func openLock(deadends []string, target string) int {
    deadendSet := make(map[string]bool)
    for _, v := range deadends {
        deadendSet[v] = true
    }
    start := "0000"
    dq := []string{start}
    seen := make(map[string]bool)
    seen[start] = true
    steps := 0
    for len(dq) > 0 {
        size := len(dq)
        for i := 0; i < size; i++ {
            curr := dq[0]
            dq = dq[1:]
            if curr == target {
                return steps
            }
            _, okDE := deadendSet[curr]
            if !okDE {
                var digit string
                for j, v := range curr {
                    if v == '9' {
                        digit = "0"
                    } else {
                        digit = string(v + 1)
                    }
                    nextNode := curr[:j] + digit + curr[j+1:]
                    _, okDE := deadendSet[nextNode]
                    _, ok := seen[nextNode]
                    if !okDE && !ok {
                        seen[nextNode] = true
                        dq = append(dq, nextNode)
                    }
                    if v == '0' {
                        digit = "9"
                    } else {
                        digit = string(v - 1)
                    }
                    nextNode = curr[:j] + digit + curr[j+1:]
                    _, okDE = deadendSet[nextNode]
                    _, ok = seen[nextNode]
                    if !okDE && !ok {
                        seen[nextNode] = true
                        dq = append(dq, nextNode)
                    }
                }
            }
        }
        steps++
    }
    return -1
}

func main() {
    type Pair struct {
        deadends []string
        target   string
    }
    arr := []Pair{
        {[]string{"0201", "0101", "0102", "1212", "2002"}, "0202"},
        {[]string{"8888"}, "0009"},
        {[]string{"8887", "8889", "8878", "8898", "8788", "8988", "7888", "9888"}, "8888"},
        {[]string{"0000"}, "8888"},
    }
    for _, stuff := range arr {
        fmt.Println(stuff.deadends, stuff.target)
        deadends, target := stuff.deadends, stuff.target
        r := openLock(deadends, target)
        fmt.Println(r)
    }
}

